# Oakland 311 calls

This notebook is part of Assignment #6.

In [1]:
import pandas as pd
import altair as alt
import requests

## Download data

Download [311 call data](https://data.oaklandca.gov/Infrastructure/Service-requests-received-by-the-Oakland-Call-Cent/quth-gb8e) from the City of Oakland's Open Data Portal. 

In [2]:
url = 'https://data.oaklandca.gov/api/views/quth-gb8e/rows.csv?accessType=DOWNLOAD'
r = requests.get(url, allow_redirects=False)

# write the content of the request into a file called `oakland_311.csv`
open('oakland_311.csv', 'wb').write(r.content)

179882250

## Read the CSV into a dataframe

Next, I read the csv just containing the 311 data into a dataframe called "oakland_311_original." I then copy that dataframe into another dataframe called "oakland_311" to use as the working copy. Finally, I call the info for the oakland_311 dataframe to see the field Dtypes and other high level summary information. 

In [3]:
#read oakland_311.csv into a dataframe called oakland_311_original
oakland_311_original = pd.read_csv('oakland_311.csv')

#copy the oakland_311_original dataframe into a dataframe called "oakland_311"
oakland_311 = oakland_311_original.copy()

#view the dataframe oakland_311
oakland_311.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 852541 entries, 0 to 852540
Data columns (total 16 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   REQUESTID        852541 non-null  int64  
 1   DATETIMEINIT     852541 non-null  object 
 2   SOURCE           852541 non-null  object 
 3   DESCRIPTION      852541 non-null  object 
 4   REQCATEGORY      852460 non-null  object 
 5   REQADDRESS       836654 non-null  object 
 6   STATUS           852540 non-null  object 
 7   REFERREDTO       70168 non-null   object 
 8   DATETIMECLOSED   610875 non-null  object 
 9   SRX              836654 non-null  float64
 10  SRY              836654 non-null  float64
 11  COUNCILDISTRICT  809790 non-null  object 
 12  BEAT             812245 non-null  object 
 13  PROBADDRESS      852064 non-null  object 
 14  City             852541 non-null  object 
 15  State            852541 non-null  object 
dtypes: float64(2), int64(1), object(13)
me

We see that there are 15 fields in this dataset. There are 852541 rows.

## df.head and df.tail

Using the .head and .tail functions, we'll take a look at the beginning and end of our dataset.

In [4]:
#get the last 5 rows of the oakland_311 dataframe
oakland_311.tail()

REQUESTID            DATETIMEINIT       SOURCE  \
852536    1151935  09/30/2021 01:54:23 PM  SeeClickFix   
852537    1151856  09/30/2021 11:26:13 AM  SeeClickFix   
852538    1147185  09/14/2021 02:54:49 PM  SeeClickFix   
852539    1150481  09/25/2021 03:31:40 PM  SeeClickFix   
852540    1151347  09/28/2021 03:54:42 PM  SeeClickFix   

                                      DESCRIPTION     REQCATEGORY  \
852536          Utility - Lid - Missing or Broken  ROW_INSPECTORS   
852537  Litter - Street Litter Container - Broken         ILLDUMP   
852538                Parking - Abandoned Vehicle          POLICE   
852539    Homeless Encampment - Living in Vehicle        HE_CLEAN   
852540    Homeless Encampment - Living in Vehicle        HE_CLEAN   

                                       REQADDRESS    STATUS REFERREDTO  \
852536    (37.80086333056584, -122.2728949987278)      OPEN        NaN   
852537     (37.7990150403045, -122.2729668018932)      OPEN        NaN   
852538   (37.78429373521214, -122.19783353979535)      OPEN        NaN   
852539   (37.75920504225326, -122.18706682059627)  UNFUNDED        NaN   
852540  (37.781792932032374, -122.24281520941372)  UNFUNDED        NaN   

       DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
852536            NaN  6049471.080  2118926.060            CCD2  03X   
852537            NaN  6049437.589  2118253.569            CCD2  03X   
852538            NaN  6071044.715  2112491.573            CCD6  28X   
852539            NaN  6073990.955  2103301.471            CCD7  30X   
852540            NaN  6058030.915  2111820.075            CCD5  19X   

                          PROBADDRESS     City State  
852536                   988 BROADWAY  Oakland    CA  
852537           7TH ST & FRANKLIN ST  Oakland    CA  
852538             3216 COURTLAND AVE  Oakland    CA  
852539  73RD AVE & INTERNATIONAL BLVD  Oakland    CA  
852540               2000 EMBARCADERO  Oakland    CA

In [5]:
#get the first 5 rows of the oakland_311 dataframe
oakland_311.head()

REQUESTID            DATETIMEINIT       SOURCE  \
0    1145885  09/09/2021 06:08:56 PM  SeeClickFix   
1    1146984  09/14/2021 08:08:50 AM        Phone   
2    1150718  09/27/2021 07:52:03 AM  SeeClickFix   
3    1151336  09/28/2021 03:37:51 PM  SeeClickFix   
4    1151681  09/29/2021 04:18:33 PM  SeeClickFix   

                                  DESCRIPTION  REQCATEGORY  \
0     Illegal Dumping (Enforcement Potential)  ENVIRON_ENF   
1     Illegal Dumping (Enforcement Potential)  ENVIRON_ENF   
2  Illegal Dumping - debris, appliances, etc.      ILLDUMP   
3                 Parking - Abandoned Vehicle       POLICE   
4                 Parking - Abandoned Vehicle       POLICE   

                                  REQADDRESS STATUS REFERREDTO DATETIMECLOSED  \
0  (37.788936670596335, -122.24357499716558)   OPEN        NaN            NaN   
1   (37.78829957727266, -122.23117705915959)   OPEN        NaN            NaN   
2   (37.82142504165362, -122.28732679694832)   OPEN        NaN            NaN   
3   (37.75306101818787, -122.16312297802386)   OPEN        NaN            NaN   
4     (37.785503590812, -122.20544605709021)   OPEN        NaN            NaN   

           SRX          SRY COUNCILDISTRICT BEAT         PROBADDRESS     City  \
0  6057859.838  2114424.905            CCD2  19X       1438 16TH AVE  Oakland   
1  6061437.527  2114126.459            CCD5  21X      2324 E 21ST ST  Oakland   
2  6045445.169  2126491.007            CCD3  07X  ETTIE ST & 28TH ST  Oakland   
3  6080871.846  2100940.775            CCD7  35X    9227 HILLSIDE ST  Oakland   
4  6068853.201  2112972.033            CCD4  24Y  3901 ALLENDALE AVE  Oakland   

  State  
0    CA  
1    CA  
2    CA  
3    CA  
4    CA

## Reformat the open and close date fields
Using pd.to.datetime, I reformat the DATETIMEINIT and DATETIMECLOSED fields to be a datetime Dtype.

In [6]:
#reformat the fields DATETIMEINIT and. DATETIMECLOSED to be a datetime Dtype

oakland_311['DATETIMEINIT'] = pd.to_datetime(oakland_311['DATETIMEINIT'])
oakland_311['DATETIMECLOSED'] = pd.to_datetime(oakland_311['DATETIMECLOSED'])

#get info on the dataframe oakland_311
oakland_311.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 852541 entries, 0 to 852540
Data columns (total 16 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   REQUESTID        852541 non-null  int64         
 1   DATETIMEINIT     852541 non-null  datetime64[ns]
 2   SOURCE           852541 non-null  object        
 3   DESCRIPTION      852541 non-null  object        
 4   REQCATEGORY      852460 non-null  object        
 5   REQADDRESS       836654 non-null  object        
 6   STATUS           852540 non-null  object        
 7   REFERREDTO       70168 non-null   object        
 8   DATETIMECLOSED   610875 non-null  datetime64[ns]
 9   SRX              836654 non-null  float64       
 10  SRY              836654 non-null  float64       
 11  COUNCILDISTRICT  809790 non-null  object        
 12  BEAT             812245 non-null  object        
 13  PROBADDRESS      852064 non-null  object        
 14  City             852

Now we'll take a look to see what the change looks like in the data by calling the head function.

In [7]:
oakland_311.head()

REQUESTID        DATETIMEINIT       SOURCE  \
0    1145885 2021-09-09 18:08:56  SeeClickFix   
1    1146984 2021-09-14 08:08:50        Phone   
2    1150718 2021-09-27 07:52:03  SeeClickFix   
3    1151336 2021-09-28 15:37:51  SeeClickFix   
4    1151681 2021-09-29 16:18:33  SeeClickFix   

                                  DESCRIPTION  REQCATEGORY  \
0     Illegal Dumping (Enforcement Potential)  ENVIRON_ENF   
1     Illegal Dumping (Enforcement Potential)  ENVIRON_ENF   
2  Illegal Dumping - debris, appliances, etc.      ILLDUMP   
3                 Parking - Abandoned Vehicle       POLICE   
4                 Parking - Abandoned Vehicle       POLICE   

                                  REQADDRESS STATUS REFERREDTO DATETIMECLOSED  \
0  (37.788936670596335, -122.24357499716558)   OPEN        NaN            NaT   
1   (37.78829957727266, -122.23117705915959)   OPEN        NaN            NaT   
2   (37.82142504165362, -122.28732679694832)   OPEN        NaN            NaT   
3   (37.75306101818787, -122.16312297802386)   OPEN        NaN            NaT   
4     (37.785503590812, -122.20544605709021)   OPEN        NaN            NaT   

           SRX          SRY COUNCILDISTRICT BEAT         PROBADDRESS     City  \
0  6057859.838  2114424.905            CCD2  19X       1438 16TH AVE  Oakland   
1  6061437.527  2114126.459            CCD5  21X      2324 E 21ST ST  Oakland   
2  6045445.169  2126491.007            CCD3  07X  ETTIE ST & 28TH ST  Oakland   
3  6080871.846  2100940.775            CCD7  35X    9227 HILLSIDE ST  Oakland   
4  6068853.201  2112972.033            CCD4  24Y  3901 ALLENDALE AVE  Oakland   

  State  
0    CA  
1    CA  
2    CA  
3    CA  
4    CA

## Add close_time field
I then add a field called "close_time" by subtracting DATETIMEINIT from DATETIMECLOSED.

In [8]:
#create a new field called "close_time" by subtracting "datetimeinit from datetimeclosed"

oakland_311['Close_Time'] = oakland_311['DATETIMECLOSED'] - oakland_311['DATETIMEINIT']

Next, I want to explore the data added to the close_time field.

## Exploring the data
Here, we use the .mean, .max. .min and .median functions to start exploring our dataset.

First, I want to find the earliest entry in the dataset as defined by the DATIMEINIT field. 

In [9]:
#using the .min function to find the earliest entry
oakland_311['DATETIMEINIT'].min()

Timestamp('2009-07-01 08:05:36')

The first entry in this dataset was on July 1, 2009.

Next, I want to find out when was the latest entry in this dataset,

In [10]:
#using the .max function to find the earliest entry
oakland_311['DATETIMEINIT'].max()

Timestamp('2021-10-30 01:29:15')

The latest entry in this dataset was on October 30, 2021.

Next, I'm going to explore the "close_time" field.

In [11]:
oakland_311['Close_Time'].mean()

Timedelta('49 days 19:45:09.623400861')

The average time Oakland took to close a case was around 49 days. This is for the whole dataset, from mid 2009 to now.

In [12]:
oakland_311['Close_Time'].max()

Timedelta('4293 days 21:59:22')

The average time Oakland took to close a case was around 49 days. This is for the whole dataset, from mid 2009 to now.

In [14]:
oakland_311['Close_Time'].median()

Timedelta('4 days 22:03:37')

The median time Oakland took to close a case was around 49 days. This is for the whole dataset, from mid 2009 to now.

In [15]:
oakland_311['Close_Time'].min()

Timedelta('-2981 days +00:37:47')

The quickest time Oakland took to close a case was around -2981 days. This shows that the data is messy and likely has errors in it. 

Next, I'll sort the data by the "close_time" field in ascending order in order to investigate how many rows might have a similar problem.

In [16]:
oakland_311.sort_values(by=['Close_Time'], ascending=True).head(20)

REQUESTID        DATETIMEINIT          SOURCE  \
251083     549531 2014-11-06 14:27:32           Phone   
226101     528272 2014-07-10 12:20:22     SeeClickFix   
208987     508652 2014-03-20 11:14:09  Phone or Email   
201412     506165 2014-03-04 14:31:02  Phone or Email   
208172     497297 2014-01-06 11:11:07           Phone   
195237     486371 2013-10-17 11:40:30           Phone   
168398     467065 2013-06-25 10:19:31  Phone or Email   
119110     403182 2012-04-03 13:20:44  Phone or Email   
13928       25209 2011-06-07 16:22:31  Phone or Email   
20794       24694 2011-03-16 14:02:57  Phone or Email   
59187      342560 2010-07-20 09:00:50  Phone or Email   
45835      340774 2010-06-29 10:46:04  Phone or Email   
31565      340773 2010-06-29 10:41:27  Phone or Email   
40612      338829 2010-06-07 15:23:49  Phone or Email   
35424      333745 2010-04-12 13:06:12  Phone or Email   
48040      332777 2010-03-29 11:21:28  Phone or Email   
37833      326467 2010-01-21 09:10:06  Phone or Email   
9144       432312 2012-11-01 11:24:56  Phone or Email   
124547     429115 2012-10-10 09:04:41         Website   
140857     427677 2012-09-28 15:11:04  Phone or Email   

                                              DESCRIPTION REQCATEGORY  \
251083               Illegal Dumping � mattress/boxspring     ILLDUMP   
226101                                Hazardous Materials        FIRE   
208987  CityBldg - Container Req (Cart/Bin from WMAC/CWS)   RECYCLING   
201412  CityBldg - Container Req (Cart/Bin from WMAC/CWS)   RECYCLING   
208172                               City Bldg - Plumbing   BLDGMAINT   
195237                 Human Waste in Public-Right-of-Way       OTHER   
168398                                Hazardous Materials        FIRE   
119110  CityBldg - Container Req (Cart/Bin from WMAC/CWS)   RECYCLING   
13928               City Services - Bin/Container Request   RECYCLING   
20794                                       City Services   RECYCLING   
59187                                   Sewers - Blockage      SEWERS   
45835                       Streets - Potholes/Depression    STREETSW   
31565                Electrical - Inside/Outside Facility   BLDGMAINT   
40612                                  Graffiti in a Park   BLDGMAINT   
35424                       Fence - City Facility or Park   BLDGMAINT   
48040                       Streets - Potholes/Depression    STREETSW   
37833                         Building - Structural Issue   BLDGMAINT   
9144                                                 PG&E       OTHER   
124547                                  City Bldg - Other   BLDGMAINT   
140857                    Litter - In Public Right of Way     ILLDUMP   

                                       REQADDRESS  STATUS  \
251083  (37.823866701293426, -122.26193144921672)  CLOSED   
226101  (37.823866701293426, -122.26193144921672)  CLOSED   
208987  (37.823866701293426, -122.26193144921672)  CLOSED   
201412  (37.823866701293426, -122.26193144921672)  CLOSED   
208172  (37.823863213586364, -122.26191484866857)  CLOSED   
195237  (37.823866701293426, -122.26193144921672)  CLOSED   
168398  (37.823866701293426, -122.26193144921672)  CLOSED   
119110    (37.823864691106074, -122.261913394719)  CLOSED   
13928    (37.82386370140158, -122.26191455551799)  CLOSED   
20794    (37.82386370140158, -122.26191455551799)  CLOSED   
59187     (37.823864691106074, -122.261913394719)  CLOSED   
45835     (37.823864691106074, -122.261913394719)  CLOSED   
31565     (37.823864691106074, -122.261913394719)  CLOSED   
40612     (37.823864691106074, -122.261913394719)  CLOSED   
35424     (37.823864691106074, -122.261913394719)  CLOSED   
48040     (37.823864691106074, -122.261913394719)  CLOSED   
37833     (37.823864691106074, -122.261913394719)  CLOSED   
9144    (37.790247170707254, -122.23014783691161)  CLOSED   
124547   (37.80123737662912, -122.26350890484095)  CLOSED   
140857   (37.78424060810989, -122.23514745044935)  CLOSE

It looks like there could be a pattern to the error - the first 17 rows here date to 2006-09-08 for the DATETIMECLOSED field. The remaining date to around the same in July 2012.

## Check for duplicates
Using the .len and .nunique functions, we'll check to see if there are duplicate records based on the "RequestID"

In [17]:
len(oakland_311)

852541

In [18]:
oakland_311['REQUESTID'].nunique()


852541

This dataset does not contain any duplicates according to this method. 

## Creating a subset of data
I want to create a subset of the oakland_311 data to show the data that appears to be malformed. 

In [19]:
oakland_311_bad = oakland_311[
    (oakland_311['Close_Time'] < '0 days')
].copy()

In [20]:
oakland_311_bad.tail()

REQUESTID        DATETIMEINIT          SOURCE  \
223078     522752 2014-06-10 11:29:36  Phone or Email   
225238     522842 2014-06-10 15:50:45  Phone or Email   
226101     528272 2014-07-10 12:20:22     SeeClickFix   
229541     522805 2014-06-10 14:02:16  Phone or Email   
251083     549531 2014-11-06 14:27:32           Phone   

                                 DESCRIPTION REQCATEGORY  \
223078  Recycling - Missed Pickup - Backyard   RECYCLING   
225238       Recycling - Repair/Replace Cart   RECYCLING   
226101                   Hazardous Materials        FIRE   
229541       Recycling - Repair/Replace Cart   RECYCLING   
251083  Illegal Dumping � mattress/boxspring     ILLDUMP   

                                       REQADDRESS  STATUS  \
223078   (37.81947378040759, -122.19864842876396)  CLOSED   
225238   (37.81389394379093, -122.25288667849271)  CLOSED   
226101  (37.823866701293426, -122.26193144921672)  CLOSED   
229541    (37.82652163613099, -122.2607234958911)  CLOSED   
251083  (37.823866701293426, -122.26193144921672)  CLOSED   

                                  REFERREDTO      DATETIMECLOSED         SRX  \
223078  CALIFORNIA WASTE SOLUTIONS, EXTERNAL 2014-06-06 16:12:25  6071041.76   
225238  CALIFORNIA WASTE SOLUTIONS, EXTERNAL 2014-06-06 16:12:25  6055339.74   
226101                 FIRE (GRIFFIN, LEROY) 2006-09-08 15:05:19  6052795.80   
229541  CALIFORNIA WASTE SOLUTIONS, EXTERNAL 2014-06-06 16:12:25  6053162.82   
251083                                   NaN 2006-09-08 15:05:19  6052795.80   

               SRY COUNCILDISTRICT BEAT        PROBADDRESS     City State  \
223078  2125303.60            CCD4  13Z  2300 MASTLANDS DR  Oakland    CA   
225238  2123561.09            CCD3  14X      302 EUCLID AV  Oakland    CA   
226101  2127240.83            CCD3  08X      MOSSWOOD PARK  Oakland    CA   
229541  2128200.82            CCD1  08X        409 38TH ST  Oakland    CA   
251083  2127240.83            CCD3  08X      MOSSWOOD PARK  Oakland    CA   

                 Close_Time  
223078    -4 days +04:42:49  
225238    -4 days +00:21:40  
226101 -2862 days +02:44:57  
229541    -4 days +02:10:09  
251083 -2981 days +00:37:47

Next, I'm going to take the request ID from the oakland_311_malformed df and store it into a new df called "bad_oakland_request_ids"

In [21]:
#put the REQUESTID in this data into a list 

bad_oakland_request_ids = oakland_311_bad['REQUESTID'].to_list()

In [22]:
bad_oakland_request_ids

[432312,
 23385,
 23396,
 24888,
 25105,
 25017,
 23399,
 24995,
 24852,
 25002,
 25000,
 24620,
 23406,
 23934,
 24856,
 317920,
 25037,
 24988,
 317005,
 25246,
 23418,
 24285,
 24664,
 23364,
 25209,
 25041,
 23391,
 24614,
 23414,
 23001,
 24617,
 24792,
 23381,
 24898,
 23376,
 24960,
 24301,
 24996,
 25413,
 23664,
 24621,
 24857,
 23386,
 23397,
 24640,
 25106,
 24297,
 24387,
 23417,
 23403,
 24853,
 23000,
 23390,
 24999,
 24794,
 25199,
 24295,
 24865,
 23413,
 24300,
 23402,
 24733,
 24645,
 23407,
 24663,
 23375,
 23394,
 23935,
 25158,
 318120,
 24850,
 25349,
 24993,
 24892,
 24369,
 25089,
 24280,
 23382,
 25206,
 24961,
 24831,
 24858,
 24848,
 25035,
 24694,
 23416,
 23383,
 23398,
 24298,
 23409,
 24622,
 23401,
 25086,
 24869,
 23374,
 23389,
 309098,
 23404,
 24662,
 22999,
 23632,
 23419,
 24849,
 23378,
 25043,
 24796,
 23393,
 25157,
 23408,
 312764,
 24661,
 25345,
 23388,
 25003,
 24994,
 24299,
 24962,
 24644,
 23412,
 24859,
 25060,
 23933,
 23410,
 25033,
 2

Next, I'll get the subset of the oakland_311 dataframe based on the negative close_time days requestids

In [23]:
#create a subset of data based on the list of request IDs I just made
oakland_311[oakland_311['REQUESTID'].isin(bad_oakland_request_ids)]

REQUESTID        DATETIMEINIT          SOURCE  \
9144       432312 2012-11-01 11:24:56  Phone or Email   
11126       23385 2010-09-16 13:59:19  Phone or Email   
11127       23396 2010-09-16 13:59:46  Phone or Email   
11226       24888 2011-04-06 16:33:06  Phone or Email   
11400       25105 2011-05-16 11:55:22  Phone or Email   
...           ...                 ...             ...   
223078     522752 2014-06-10 11:29:36  Phone or Email   
225238     522842 2014-06-10 15:50:45  Phone or Email   
226101     528272 2014-07-10 12:20:22     SeeClickFix   
229541     522805 2014-06-10 14:02:16  Phone or Email   
251083     549531 2014-11-06 14:27:32           Phone   

                                            DESCRIPTION REQCATEGORY  \
9144                                               PG&E       OTHER   
11126              Residential Recycling Service Issues   RECYCLING   
11127              Residential Recycling Service Issues   RECYCLING   
11226   Garbage - Requests Exemption - Weekly or Shared   RECYCLING   
11400              Residential Recycling Service Issues   RECYCLING   
...                                                 ...         ...   
223078             Recycling - Missed Pickup - Backyard   RECYCLING   
225238                  Recycling - Repair/Replace Cart   RECYCLING   
226101                              Hazardous Materials        FIRE   
229541                  Recycling - Repair/Replace Cart   RECYCLING   
251083             Illegal Dumping � mattress/boxspring     ILLDUMP   

                                       REQADDRESS  STATUS  \
9144    (37.790247170707254, -122.23014783691161)  CLOSED   
11126   (30.009899650831787, -141.21875098935422)  CLOSED   
11127   (30.009899650831787, -141.21875098935422)  CLOSED   
11226   (30.009899650831787, -141.21875098935422)  CLOSED   
11400   (30.009899650831787, -141.21875098935422)  CLOSED   
...                                           ...     ...   
223078   (37.81947378040759, -122.19864842876396)  CLOSED   
225238   (37.81389394379093, -122.25288667849271)  CLOSED   
226101  (37.823866701293426, -122.26193144921672)  CLOSED   
229541    (37.82652163613099, -122.2607234958911)  CLOSED   
251083  (37.823866701293426, -122.26193144921672)  CLOSED   

                                  REFERREDTO      DATETIMECLOSED         SRX  \
9144                                   PG&E, 2012-07-10 14:00:14  6061748.00   
11126                                    NaN 2010-09-16 13:59:07        0.00   
11127                                    NaN 2010-09-16 13:59:35        0.00   
11226                                    NaN 2011-04-06 16:33:04        0.00   
11400                                    NaN 2011-05-16 11:55:21        0.00   
...                                      ...                 ...         ...   
223078  CALIFORNIA WASTE SOLUTIONS, EXTERNAL 2014-06-06 16:12:25  6071041.76   
225238  CALIFORNIA WASTE SOLUTIONS, EXTERNAL 2014-06-06 16:12:25  6055339.74   
226101                 FIRE (GRIFFIN, LEROY) 2006-09-08 15:05:19  6052795.80   
229541  CALIFORNIA WASTE SOLUTIONS, EXTERNAL 2014-06-06 16:12:25  6053162.82   
251083                                   NaN 2006-09-08 15:05:19  6052795.80   

               SRY COUNCILDISTRICT BEAT          PROBADDRESS     City State  \
9144    2114830.00             NaN  20X  E 23RD ST & 23RD AV  Oakland    CA   
11126         0.00             NaN  NaN                   ZZ  Oakland    CA   
11127         0.00             NaN  NaN                   ZZ  Oakland    CA   
11226         0.00             NaN  NaN                   ZZ  Oakland    CA   
11400         0.00             NaN  NaN                   ZZ  Oakland    CA   
...            ...             ...  ...                  ...      ...   ...   
223078  2125303.60            CCD4  13Z    2300 MASTLANDS DR  Oakland    CA   
225238  2123561.09            CCD3  14X        302 EUCLID AV  Oakland    CA   
226101  2127240.83            CCD3  08X        MOSSWOOD PARK  Oak

## Creating a clean dataframe without negative clean_time entries
Next, I'll create a clean dataframe by filtering out rows with a close_time less than 0

In [24]:
oakland_311_clean = oakland_311[
    (oakland_311['Close_Time'] > '0 days')
].copy()

In [25]:
#show the dataframe with the bad data removed
oakland_311_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 606292 entries, 7 to 852535
Data columns (total 17 columns):
 #   Column           Non-Null Count   Dtype          
---  ------           --------------   -----          
 0   REQUESTID        606292 non-null  int64          
 1   DATETIMEINIT     606292 non-null  datetime64[ns] 
 2   SOURCE           606292 non-null  object         
 3   DESCRIPTION      606292 non-null  object         
 4   REQCATEGORY      606256 non-null  object         
 5   REQADDRESS       592159 non-null  object         
 6   STATUS           606292 non-null  object         
 7   REFERREDTO       2440 non-null    object         
 8   DATETIMECLOSED   606292 non-null  datetime64[ns] 
 9   SRX              592159 non-null  float64        
 10  SRY              592159 non-null  float64        
 11  COUNCILDISTRICT  584746 non-null  object         
 12  BEAT             586460 non-null  object         
 13  PROBADDRESS      605968 non-null  object         
 14  City

We now have a dataset that excludes negative values and NaT values. 

## Export to CSV
Export the cleaned dataframe to a csv called 'oakland_311_cleaned.csv'

In [27]:
oakland_311_clean.to_csv('oakland_311.csv', index=False)